projects/plugins/jetpack/class.json-api-endpoints.php

```diff
@@ -2618,6 +2649,161 @@ public function get_amp_cache_origins( $siteurl ) {
 		);
 	}
 
+	/**
+	 * Register a REST route for this jsonAPI endpoint.
+	 *
+	 * @return void
+	 * @throws Exception The exception if something goes wrong.
+	 */
+	public function create_rest_route_for_endpoint() {
+		register_rest_route(
+			static::REST_NAMESPACE,
+			$this->build_rest_route(),
+			array(
+				'methods'             => $this->method,
+				'callback'            => array( $this, 'rest_callback' ),
+				'permission_callback' => array( $this, 'rest_permission_callback' ),
+			)
+		);
+	}
+
+	/**
+	 * Handle the rest call.
+	 *
+	 * @param WP_REST_Request $request The request object.
+	 *
+	 * @return mixed|WP_Error
+	 */
+	public function rest_callback( WP_REST_Request $request ) {
+		// phpcs:ignore WordPress.PHP.IniSet.display_errors_Disallowed -- Making sure random warnings don't break JSON.
+		ini_set( 'display_errors', false );
+
+		$blog_id = Jetpack_Options::get_option( 'id' );
+
+		$this->api->initialize();
+		$this->api->endpoint = $this;
+
+		$locale = $request->get_param( 'language' );
+		if ( $locale ) {
+			$this->api->init_locale( $locale );
+		}
+
+		if ( $this->in_testing && ! WPCOM_JSON_API__DEBUG ) {
+			return new WP_Error( 'endpoint_not_available' );
+		}
+
+		/** This action is documented in class.json-api.php */
+		do_action( 'wpcom_json_api_output', $this->stat );
+
+		$callback_response = call_user_func_array(
+			array( $this, 'callback' ),
+			array_values( array( $this->path, $blog_id ) + $request->get_url_params() )
+		);
+
+		if ( ! $callback_response && ! is_array( $callback_response ) ) {
```

Comment:
Do we really need to use the `output` method? Mostly asking because it seems to overcomplicate the approach.. 
I was testing this out and it seems to work fine if we simply return early in case of `WP_Error`: `if ( is_wp_error( $callback_response ) ) {` and leave the rest logic same as before.

Reply:
Hi @fgiannar 👋 
We use the `output()` method here to wrap the response if `http_envelope=1` is provided.
There was a bug though which didn't make it all work properly, fixed in https://github.com/Automattic/jetpack/pull/39432/commits/bbb226a966c98d9076c2f07b2ed3307236a1d03f.

Here's what I get when I make `$callback_response = null;` with `http_envelope`:
```
HTTP/2 200
{"code":400,"headers":[{"name":"Content-Type","value":"application\/json"}],"body":{"error":"server_error","message":"The Jetpack site is inaccessible or returned an error: transport error - HTTP status code was not 200 (500) [-32300]"}}
```

And without `http_envelope`:
```
HTTP/2 400
{"error":"server_error","message":"The Jetpack site is inaccessible or returned an error: transport error - HTTP status code was not 200 (500) [-32300]"}
```